src/Disks/ObjectStorages/MetadataStorageFromObject.cpp

```diff
@@ -0,0 +1,219 @@
+}
+
+const std::string & MetadataStorageFromObject::getPath() const
+{
+    return object_storage_root_path;
+}
+
+bool MetadataStorageFromObject::exists(const std::string & path) const
+{
+    auto object = StoredObject::create(*object_storage, fs::path(object_storage_root_path) / path);
+    return object_storage->exists(object);
+}
+
+bool MetadataStorageFromObject::isFile(const std::string &) const
+{
+    return true;
+}
+
+bool MetadataStorageFromObject::isDirectory(const std::string &) const
+{
+    return false;
+}
+
+Poco::Timestamp MetadataStorageFromObject::getLastModified(const std::string &) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "getLastModified is not implemented for MetadataStorageFromObject");
+}
+
+struct stat MetadataStorageFromObject::stat(const std::string &) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "stat is not implemented for MetadataStorageFromObject");
+}
+
+time_t MetadataStorageFromObject::getLastChanged(const std::string &) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "getLastChanged is not implemented for MetadataStorageFromObject");
+}
+
+uint64_t MetadataStorageFromObject::getFileSize(const String & path) const
+{
+    RelativePathsWithSize children;
+    object_storage->listPrefix(path, children);
+    if (children.empty())
+        return 0;
+    if (children.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "listPrefix() return multiple paths ({}) for {}", children.size(), path);
+    return children.front().bytes_size;
+}
+
+std::vector<std::string> MetadataStorageFromObject::listDirectory(const std::string & path) const
+{
+    RelativePathsWithSize children;
+    object_storage->listPrefix(path, children);
+
+    std::vector<std::string> result;
+    for (const auto & path_size : children)
+    {
+        result.push_back(path_size.relative_path);
+    }
+    return result;
+}
+
+DirectoryIteratorPtr MetadataStorageFromObject::iterateDirectory(const std::string & path) const
+{
+    /// NOTE: this is not required for BACKUP/RESTORE, but this is a first step
+    /// towards MergeTree on plain S3.
+    auto paths = listDirectory(path);
+    std::vector<std::filesystem::path> fs_paths(paths.begin(), paths.end());
+    return std::make_unique<StaticDirectoryIterator>(std::move(fs_paths));
+}
+
+std::string MetadataStorageFromObject::readFileToString(const std::string &) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "readFileToString is not implemented for MetadataStorageFromObject");
+}
+
+std::unordered_map<String, String> MetadataStorageFromObject::getSerializedMetadata(const std::vector<String> &) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "getSerializedMetadata is not implemented for MetadataStorageFromObject");
+}
+
+StoredObjects MetadataStorageFromObject::getStorageObjects(const std::string & path) const
+{
+    std::string blob_name = object_storage->generateBlobNameForPath(path);
```

Comment:
Is it just path? Can't we write `blob_name = path` here?

Reply:
Yes, but it cannot use `S3ObjectStorage` anyway, since `generateBlobNameForPath` used in other places, so I prefer to make it consistent and use `generateBlobNameForPath` here too. First I put a comment that it is OK, but later decided that it is redundant. Thoughts?

Reply:
Which places?

Reply:
`DiskObjectStorageTransaction`